interop-java/src/main/java/scalaz/zio/java/IO.java

```diff
@@ -27,9 +28,21 @@ private IO(scalaz.zio.IO<E, A> delegate) {
         return new IO<>(delegate.flatMap(f.andThen(io -> io.delegate)::apply));
     }
 
-    // TODO Fiber
-    // final def fork: IO[Nothing, Fiber[E, A]]
-    // final def fork0(handler: Cause[Any] => IO[Nothing, Unit]): IO[Nothing, Fiber[E, A]]
+    // TODO do we want a separate java Fiber?
+    public IO<Void, Fiber<E, A>> fork() {
+        return new IO<>(delegate.fork()).leftMap(nothing -> null);
+    }
+
+    public IO<Void, Fiber<E, A>> fork0(Function<ExitResult.Cause<Object>, IO<Void, Void>> f) {
+        return new IO<>(
+                delegate.fork0(f.andThen(io ->
+                        io.delegate.<Object, BoxedUnit>bimap(
+                                _void -> _void,
+                                _void -> BoxedUnit.UNIT
+                        ))::apply
+                )
+        ).bimap(nothing -> null, nothing -> null);
```

Comment:
Akka is using `NotUsed` and `Done` thing,should we introduce those like things for Java API too, or just `Void`?

Reply:
Just `Void` is sufficient. However, we will need helpers:

```java
public static <E, A> IO<E, A> asError(IO<Void, A> io) { ...}
public static <E, A> IO<E, A> asValue(IO<E, Void> io) { ...}
```

In Scala, we get these "up conversions" for free, because of declaration site covariance. In Java, users will have to explicitly "up cast".

Reply:
Oh, by the way, these have an error: `Void` should only be used in place of `Nothing`, **not** `Unit`.

For `Unit`, we probably need to add this:

```java
class Unit {
  private Unit() { }

  public static Unit Unit = new Unit();
}
```

There is one value of type `Unit`, and zero values of type `Void` (/`Nothing`).

Reply:
I actually thought Java provided autoboxing from `void` into `Void`, which would make it similar to Scala's Unit. Since it doesn't, I'll create a `Unit`. However, I think we should come up with a more intuitive name for it.

Reply:
In the reflection API, `Void` represents the type of a method returning `void`. However, you can't return a value of type `Void`, rather, you have to return `null`.

I'm not sure this exactly corresponds to the meaning of `Nothing`, but no matter what type we created (e.g. `class Nothing { private Nothing() }`), it's still going to be possible to use `null` for values of that type, due to the Java type system. I don't think we should consider `null` a valid value.

Reply:
Akka introduced `NotUsed` and `Done` for this kind of thing.

Reply:
I have to say, sometimes using `Void` in Java is a really bad idea. Like what I just did in https://github.com/akka/akka/pull/26024

Reply:
@jdegoes or something like this, right?
```
public enum Unit {
    VALUE
}
```